### PR TITLE
Allow programmatic configuration of unicast relays.

### PR DESCRIPTION
**Pull request by mbeards**
This change allows users to configure relays from code without having to `setenv(GZ_RELAY)`.

# 🎉 New feature

## Summary
This change allows users to configure relays from code without having to `setenv(GZ_RELAY)`.

Supersedes https://github.com/gazebosim/gz-transport/pull/497

## Test it
This can be tested by starting a Node across a boundary where multicast isn't enabled, and setting the relay IP via NodeOptions.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by mbeards** [comment]
Friendly ping @caguero

**Review by caguero — COMMENTED** [review]
By making `Discovery::AddRelayAddress()` public, I think we now have risk for potential race conditions, as `this->relayAddrs` can be modified from a user thread (via `Node`) and the discovery thread can read the same variable via `SendUnicast()`.

**Comment by mbeards** [comment]
> By making `Discovery::AddRelayAddress()` public, I think we now have risk for potential race conditions, as `this->relayAddrs` can be modified from a user thread (via `Node`) and the discovery thread can read the same variable via `SendUnicast()`.

Should I add a new mutex to guard relayAddrs or just use the existing `Discovery::mutex`?

**Comment by caguero** [comment]
> > By making `Discovery::AddRelayAddress()` public, I think we now have risk for potential race conditions, as `this->relayAddrs` can be modified from a user thread (via `Node`) and the discovery thread can read the same variable via `SendUnicast()`.
> 
> Should I add a new mutex to guard relayAddrs or just use the existing `Discovery::mutex`?

I think you can reuse `Discovery::mutex`.

**Comment by mbeards** [comment]
> > > By making `Discovery::AddRelayAddress()` public, I think we now have risk for potential race conditions, as `this->relayAddrs` can be modified from a user thread (via `Node`) and the discovery thread can read the same variable via `SendUnicast()`.
> > 
> > 
> > Should I add a new mutex to guard relayAddrs or just use the existing `Discovery::mutex`?
> 
> I think you can reuse `Discovery::mutex`.

Is there a performance concern to worry about re: serializing the user thread + discovery thread in the presence of unicast relays? Every unicast write would need to acquire that mutex.

**Comment by mbeards** [comment]
Thoughts on this test failure https://build.osrfoundation.org/job/gz_transport-pr-win/99/ ? I'm having a hard time figuring out how this PR would trigger behavior change on those tests.

**Comment by caguero** [comment]
> > > > By making `Discovery::AddRelayAddress()` public, I think we now have risk for potential race conditions, as `this->relayAddrs` can be modified from a user thread (via `Node`) and the discovery thread can read the same variable via `SendUnicast()`.
> > > 
> > > 
> > > Should I add a new mutex to guard relayAddrs or just use the existing `Discovery::mutex`?
> > 
> > 
> > I think you can reuse `Discovery::mutex`.
> 
> Is there a performance concern to worry about re: serializing the user thread + discovery thread in the presence of unicast relays? Every unicast write would need to acquire that mutex.



> > > > By making `Discovery::AddRelayAddress()` public, I think we now have risk for potential race conditions, as `this->relayAddrs` can be modified from a user thread (via `Node`) and the discovery thread can read the same variable via `SendUnicast()`.
> > > 
> > > 
> > > Should I add a new mutex to guard relayAddrs or just use the existing `Discovery::mutex`?
> > 
> > 
> > I think you can reuse `Discovery::mutex`.
> 
> Is there a performance concern to worry about re: serializing the user thread + discovery thread in the presence of unicast relays? Every unicast write would need to acquire that mutex.

I'd expect a bit of impact if unicast is used because the mutex will be acquired when calling `DispacthDiscoveryMsg()`. It shouldn't make a big difference.

**Comment by caguero** [comment]
> Thoughts on this test failure https://build.osrfoundation.org/job/gz_transport-pr-win/99/ ? I'm having a hard time figuring out how this PR would trigger behavior change on those tests.

`ReplayTest` looks unrelated.

File: include/gz/transport/Discovery.hh [hunk 1]
```diff
@@ -1535,6 +1554,9 @@ namespace gz
       /// \brief Collection of socket addresses used as remote relays.
       private: std::vector<sockaddr_in> relayAddrs;
 
+      /// \brief Mutex to guarantee exclusive write access to relayAddrs.
+      private: mutable std::shared_mutex relayAddrsMutex;
```
Comment (caguero) [inline include/gz/transport/Discovery.hh:1558]: Why do you prefer to create a new mutex?

Comment (mbeards) [reply include/gz/transport/Discovery.hh:1558]: Just to get to narrower locking. I can change it back to use Discovery::mutex if you'd prefer.

Comment (caguero) [reply include/gz/transport/Discovery.hh:1558]: I think so, having multiple mutexes always makes me a bit more nervous as it's a bit harder to reason about deadlocks.

Comment (mbeards) [reply include/gz/transport/Discovery.hh]: Done.

**Comment by mbeards** [comment]
Friendly ping @caguero .

**Review by caguero — COMMENTED** [review]
Minor comments.

**Comment by mbeards** [comment]
> Minor comments.

Done.

**Review by caguero — APPROVED** [review]

**Review by caguero — COMMENTED** [review]
Not sure why the ABI checker doesn't catch it but I think that changing the access right (from private to public in our case) in `Discovery::AddRelayAddress()` breaks ABI.

I think the solution is to create a new public version of the function that calls the existing one.

File: src/Node_TEST.cc [hunk 2]
```diff
@@ -2341,6 +2341,31 @@ TEST(NodeTest, statistics)
   EXPECT_EQ(std::nullopt, node.TopicStats("/test"));
 }
 
+//////////////////////////////////////////////////
+/// \brief Test adding and querying relays
+TEST(NodeTest, relay) {
+  transport::Node node;
+
+  // Make sure the relay we're about to add hasn't already been configured.
+  const std::vector<std::string> relays_before_add = node.GlobalRelays();
```
Comment (caguero) [inline src/Node_TEST.cc:2350]: Nit: CamelCase all variables.

**Comment by mbeards** [comment]
> Not sure why the ABI checker doesn't catch it but I think that changing the access right (from private to public in our case) in `Discovery::AddRelayAddress()` breaks ABI.
> 
> I think the solution is to create a new public version of the function that calls the existing one.

Let's just target main?

**Comment by mjcarroll** [comment]
> Not sure why the ABI checker doesn't catch it but I think that changing the access right (from private to public in our case) in Discovery::AddRelayAddress() breaks ABI.

I think it's actually fine on Linux, it's only an issue with MSVC where the symbol mangling includes the access rights

**Comment by caguero** [comment]
@osrf-jenkins run tests

**Review by caguero — APPROVED** [review]

**Comment by caguero** [comment]
Not sure why this `gz_transport-ci-pr_any-jammy-amd64` job is stuck.

**Comment by azeey** [comment]
Our CI was recently changed so that `main` branches only running tests on Noble. But the branch protection for gz-transport wasn't updated. Should be good now.

**Comment by azeey** [comment]
We do need to update the GitHub workflow to use Noble . @caguero would you mind doing that when you get a chance?

**Comment by caguero** [comment]
> We do need to update the GitHub workflow to use Noble . @caguero would you mind doing that when you get a chance?

Thanks @scpeters https://github.com/gazebosim/gz-transport/pull/504

**Comment by iche033** [comment]
@mergifyio backport gz-transport13

**Pull request merged**